source/robot_model/src/Model.cpp

```diff
@@ -260,9 +262,138 @@ std::vector<state_representation::CartesianPose> Model::forward_geometry(const s
+      Psi[i] = this->robot_model_.upperPositionLimit[i] - margin - std::max(q[i], this->robot_model_.lowerPositionLimit[i]);
+    }
+    else if(this->robot_model_.upperPositionLimit[i] - margin < q[i]) {
+      Psi[i] = this->robot_model_.lowerPositionLimit[i] + margin - std::min(q[i], this->robot_model_.upperPositionLimit[i]);
+    }     
+  }
+  
+  return Psi;
+}
+
+void Model::clamp_joint_positions(state_representation::JointPositions& joint_positions) {
+  Eigen::VectorXd q_clamped = joint_positions.get_positions();
+  for(int i = 0; i < this->robot_model_.nq; i++) {
+    q_clamped[i] = std::min(std::max(q_clamped[i], this->robot_model_.lowerPositionLimit[i]), this->robot_model_.upperPositionLimit[i]);
+  }
+  joint_positions.set_positions(q_clamped);
+}
+
+state_representation::JointPositions Model::inverse_geometry(const state_representation::CartesianState& desired_cartesian_state,
+                                                            const state_representation::JointState& current_joint_state,
+                                                            std::string frame_name,
+                                                            Model::inverse_geometry_parameters params) {
+
+  if (frame_name.empty()) {
+    // get last frame if none specified
+    frame_name = this->robot_model_.frames.back().name;
+  }
+  else {
+    // throw error if specified frame does not exist
+    if (!this->robot_model_.existFrame(frame_name)) {
+      throw (exceptions::FrameNotFoundException(frame_name));
+    }
+  }
+
+  state_representation::JointPositions q(current_joint_state);
+  state_representation::JointVelocities dq(this->get_robot_name(), this->get_number_of_joints());
+
+  const int frame_id = this->robot_model_.getFrameId(frame_name);
+  
+  const pinocchio::SE3 oMdes(desired_cartesian_state.get_orientation().toRotationMatrix(), desired_cartesian_state.get_position());
+  
+  // 1 second for the Newton-Raphson method
+  std::chrono::nanoseconds dt((int)(1e9));
+
+  pinocchio::Data::Matrix6x J(6,this->robot_model_.nq);
+  pinocchio::Data::Matrix6x J_b(6,this->robot_model_.nq);
+  J.setZero();
+  J_b.setZero();
+  Eigen::VectorXd v(this->robot_model_.nq);
+  Eigen::VectorXd delta_q(this->robot_model_.nq);
+  Eigen::MatrixXd W_b = Eigen::MatrixXd::Identity(this->robot_model_.nq, this->robot_model_.nq);
+  Eigen::MatrixXd W_c = Eigen::MatrixXd::Identity(this->robot_model_.nq, this->robot_model_.nq);
+  Eigen::VectorXd psi(this->robot_model_.nq);
+
+  this->forward_geometry(q, frame_id);
+    
+  // error in SE(3)
+  const pinocchio::SE3 dMf = oMdes.actInv(this->robot_data_.oMf[frame_id]);
+  typedef Eigen::Matrix<double, 6, 1> Vector6d;
+  Vector6d err = pinocchio::log6(dMf).toVector();
+  
+  int i=0;
+  while((err.array().abs().maxCoeff() > params.tolerance) && (i < params.max_number_of_iterations))
+  {
+    pinocchio::computeFrameJacobian(this->robot_model_,this->robot_data_,q.get_positions(),frame_id, J);
```

Comment:
- space between input arguments
```suggestion
    pinocchio::computeFrameJacobian(this->robot_model_, this->robot_data_, q.get_positions(), frame_id, J);
```